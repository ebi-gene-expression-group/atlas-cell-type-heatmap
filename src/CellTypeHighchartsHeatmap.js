import React from "react"
import PropTypes from 'prop-types'

import Highcharts from "highcharts"
import HighchartsReact from "highcharts-react-official"
import HC_heatmap from "highcharts/modules/heatmap"
import HighchartsExporting from 'highcharts/modules/exporting'
import HighchartsExportData from 'highcharts/modules/export-data'

// init the module
async function addModules() {
  HC_heatmap(Highcharts)
  HighchartsExporting(Highcharts)
  HighchartsExportData(Highcharts)
}

addModules()

Highcharts.SVGRenderer.prototype.symbols.download = (x, y, w, h) => [
  // Arrow stem
  `M`, x + w * 0.5, y,
  `L`, x + w * 0.5, y + h * 0.7,
  // Arrow head
  `M`, x + w * 0.3, y + h * 0.5,
  `L`, x + w * 0.5, y + h * 0.7,
  `L`, x + w * 0.7, y + h * 0.5,
  // Box
  `M`, x, y + h * 0.9,
  `L`, x, y + h,
  `L`, x + w, y + h,
  `L`, x + w, y + h * 0.9
]

const CellTypeHighchartsHeatmap = props => {
  const { chartHeight, hasDynamicHeight, heatmapRowHeight, axisData, heatmapData } = props

  let matrixData = []
  let i=0, j=0
  if(axisData.x && heatmapData) {
    while(i < axisData.x.length){
      while(j < axisData.y.length){
        if(heatmapData[j][i] === 0) {matrixData.push([i,j,null])}
        else if(heatmapData[j][i]!==undefined){matrixData.push([i, j, Math.floor(heatmapData[j][i] * 1000) / 1000 + 1])}
        else{matrixData.push([i,j,null])}
        j++
      }
      i++
      j=0
    }
  }

  const options = {
    chart: {
      type: `heatmap`,
      zoomType: `y`,
      height: hasDynamicHeight ? axisData.y && axisData.y.length * heatmapRowHeight : chartHeight,
      animation: false,
      marginRight: matrixData.length !== 0 ? 100 : 0,
      plotBackgroundColor: `#eaeaea`,
      spacingBottom: 0
    },
    lang: {
      noData: `There are no marker genes for this k value. Try selecting another k.`,
    },
    noData: {
      style: {
        fontWeight: `bold`,
        fontSize: `16px`,
        color: `#000000`
      }
    },
    credits: {
      enabled: false
    },
    labels: {
      step: 1
    },
    title: {
      text: `Heatmap for cell types vs marker genes`
    },

    xAxis: {
      categories: axisData.x,
      opposite: true,
      labels: {
        autoRotation: [-90]}
    },

    yAxis: {
      categories: axisData.y,
      credits: {
        enabled: false
      },
      title: false,
      reversed: true,
      startOnTick: false,
      endOnTick: false,
      gridLineWidth: 0,
      minorGridLineWidth: 0,
      showEmpty: false
    },

    tooltip: {
      formatter: function () {
        if(this.point.value === null) {
          return `<b>Cell Type:</b> ${this.series.xAxis.categories[this.point.x]} <br/> 
                  <b>Gene ID:</b> ${this.series.yAxis.categories[this.point.y]} <br /> 
                  <b>Average expression:</b> Not expressed <br/>`
        }
        else {
          return `<b>Cell Type:</b> ${this.series.xAxis.categories[this.point.x]} <br/> 
                  <b>Gene ID:</b> ${this.series.yAxis.categories[this.point.y]} <br /> 
                  <b>Average expression:</b> ${Math.round(this.point.value-1)} CPM <br/>`
        }
      }
    },

    colorAxis: {
      type: `logarithmic`,
      min: 0.1,
      max: 100000,
      stops: [
        [0, `#ffffff`],
        [0.67, `#6077bf`],
        [1, `#0e0573`]
      ],
      marker: {
        color: `#e96b23`
      }
    },

    legend: {
      title: {
        text: `Median expression (CPM)`
      },
      align: `center`,
      verticalAlign: `top`,
      layout: `horizontal`,
      symbolWidth: 480,
      enabled: matrixData.length !== 0
    },

    series: [
      {
        data: matrixData,
        nullColor: `#eaeaea`,
        cursor: `crosshair`,
        states: {
          hover: {
            brightness: 0,
            borderWidth: 2,
            borderColor: `#e96b23`
          }
        },
        turboThreshold: 0
      }
    ],
    boost: {
      useGPUTranslations: true
    },
    navigation: {
      buttonOptions: {
        theme: {
          style: {
            fontSize: `15px`
          },
          states: {
            select: {
              style: {
                fontWeight: `normal`,
                color: `black`
              }
            }
          }
        }
      },
      menuItemStyle: {
        fontSize: `15px`
      }
    },

    exporting: {
      buttons: {
        contextButton: {
          text: `Download`,
          symbol: `download`,
          menuItems: [
            `printChart`,
            `separator`,
            `downloadPNG`,
            `downloadJPEG`,
            `downloadPDF`,
            `downloadSVG`,
            `separator`,
            `downloadCSV`,
            `downloadXLS`
          ]
        }
      }
    }
  }

  return <HighchartsReact highcharts={Highcharts} options={options} />
}

CellTypeHighchartsHeatmap.propTypes = {
  chartHeight: PropTypes.number.isRequired,
  hasDynamicHeight: PropTypes.bool.isRequired,
  heatmapRowHeight: PropTypes.number.isRequired,
  axisData: PropTypes.shape({
    y: PropTypes.array,
    x: PropTypes.array
  }).isRequired,
  heatmapData: PropTypes.array.isRequired
}

export default CellTypeHighchartsHeatmap
