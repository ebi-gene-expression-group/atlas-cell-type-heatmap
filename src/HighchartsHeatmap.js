import React from "react"
import PropTypes from 'prop-types'

import Highcharts from "highcharts"
import HighchartsReact from "highcharts-react-official"
import HC_heatmap from "highcharts/modules/heatmap"

// init the module
HC_heatmap(Highcharts)

const HighchartsHeatmap = props => {
  let matrixData = []
  let i=0, j=0
  if(props.axisData.x && props.heatmapData) {
    while(i < props.axisData.x.length){
      while(j < props.axisData.y.length){
        matrixData.push([i, j, Math.floor(props.heatmapData[j][i] * 1000) / 1000 + 1])
        j++
      }
      i++
      j=0
    }
  }

  const options = {
    chart: {
      type: `heatmap`,
      zoomType: `y`,
      height: props.axisData.y && props.axisData.y.length*20,
      width: props.axisData.x && props.axisData.x.length*100

    },
    credits: {
      enabled: false
    },
    labels: {
      step: 1
    },
    title: {
      text: `Heatmap for cell types vs marker genes`
    },

    xAxis: {
      categories: props.axisData.x,
      opposite: true,
      labels: {
        autoRotation: [-20]}
    },

    yAxis: {
      categories: props.axisData.y,
      credits: {
        enabled: false
      },
      title: false,
      reversed: true,
      startOnTick: false,
      endOnTick: false,
      gridLineWidth: 0,
      minorGridLineWidth: 0,
      showEmpty: false
    },

    tooltip: {
      formatter: function () {
        if(this.point.value === null) {
          return `<b>Cell Type:</b> ${this.series.xAxis.categories[this.point.x]} <br/> 
                  <b>Gene ID:</b> ${this.series.yAxis.categories[this.point.y]} <br /> 
                  <b>Average expression:</b> Not expressed <br/>`
        }
        else {
          return `<b>Cell Type:</b> ${this.series.xAxis.categories[this.point.x]} <br/> 
                  <b>Gene ID:</b> ${this.series.yAxis.categories[this.point.y]} <br /> 
                  <b>Average expression:</b> ${Math.round(this.point.value-1)} CPM <br/>`
        }
      }
    },

    colorAxis: {
      type: `logarithmic`,
      minColor: `#FFFFFF`,
      maxColor: `#024990`,
      stops: [[0, `#FFFFFF`], [0.67, `#9ecae1`], [1, `#3182bd`]],
    },

    legend: {
      align: `center`,
      verticalAlign: `top`,
      layout: `horizontal`,
      symbolWidth: 280
    },

    series: [
      {
        data: matrixData,
        nullColor: `#eaeaea`,
        states: {
          hover: {
            brightness: 0,
            borderColor: `coral`,
            borderWidth: 2
          }
        },
        borderWidth: 1,
        borderColor: `#dddddd`
      }
    ]
  }

  return <HighchartsReact highcharts={Highcharts} options={options} />
}

HighchartsHeatmap.propTypes = {
  axisData: PropTypes.shape({
    y: PropTypes.array,
    x: PropTypes.array
  }).isRequired,
  heatmapData: PropTypes.array.isRequired
}

export default HighchartsHeatmap
