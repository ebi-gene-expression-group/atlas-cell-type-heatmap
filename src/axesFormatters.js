import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import ReactDOMServer from 'react-dom/server'

import escapedHtmlDecoder from 'he'

const reactToHtml = component => escapedHtmlDecoder.decode(ReactDOMServer.renderToStaticMarkup(component))

const ExperimentIconDiv = styled.div`
  background-color: ${props => props.background};
  color: ${props => props.color};
  border-radius: 50%;
  font-size: 16px;
  height: 20px;
  width: 20px;
  text-align: center;
  padding-top: 1px;
  vertical-align: middle;
  margin-right: 6px;
  opacity: 0.4;
  display: inline-block;
`

const XAxisLabel = ({experimentAccession, technologyType}) => {
  const technologyTypeIcon = [ `10xV2`, `Drop-seq`, `10xV3`].includes(technologyType) ?
    <ExperimentIconDiv key={experimentAccession} background={`green`} color={`white`} data-toggle={`tooltip`} data-placement={`bottom`}
      title={`Droplet CPM`}>C</ExperimentIconDiv> :
    <ExperimentIconDiv key={experimentAccession} background={`orangered`} color={`white`} data-toggle={`tooltip`} data-placement={`bottom`}
      title={`Smart-like TPM`}>T</ExperimentIconDiv>

  const geneNameWithLink =
    <a href={`http://www.ebi.ac.uk/gxa/sc/experiments/${experimentAccession}`} style={{border: `none`, color: `#148ff3`}}>
      {[technologyTypeIcon, experimentAccession]}
    </a>

  return (
    <span title={experimentAccession}>
      {geneNameWithLink}
    </span>
  )
}

const YAxisLabel = ({species, geneSymbol}) => {
  const geneSearchResultUrl = `https://www.ebi.ac.uk/gxa/sc/search?q=${geneSymbol}&species=${species}`
  return <a href={geneSearchResultUrl} style={{border: `none`, color: `#148ff3`}}>
    {geneSymbol}
  </a>
}

XAxisLabel.propTypes = {
  experimentAccession: PropTypes.string.isRequired,
  technologyType: PropTypes.string.isRequired
}

YAxisLabel.propTypes = {
  species: PropTypes.string.isRequired,
  geneSymbol: PropTypes.string.isRequired
}


const XAxisFormatter = value => reactToHtml(
  <XAxisLabel
    experimentAccession={value.experimentAccession}
    technologyType={value.technologyType}
  />
)

const YAxisFormatter = (species, geneSymbol) => {
  return reactToHtml(
    <YAxisLabel
      species={species.species}
      geneSymbol={geneSymbol}
    />
  )
}

export {XAxisFormatter, YAxisFormatter}
