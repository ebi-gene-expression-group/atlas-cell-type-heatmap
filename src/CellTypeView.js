import React from "react"
import PropTypes from 'prop-types'

import URI from 'urijs'
import _ from "lodash"

import CellTypeHighchartsHeatmap from "./CellTypeHighchartsHeatmap"
import CalloutAlert from './CalloutAlert'
import LoadingOverlay from './LoadingOverlay'
import DataSeriesHeatmapLegent from './DataSeriesHeatmapLegend'


class CellTypeView extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      expressionByCellTypeData: [],
      solrResponse: [],
      filteredData: [],
      experimentAccessionsBySpecies: [],
      responseBySpecies: [],
      axisData: {},
      isLoading: false,
      speciesList: [],
      selectedSpecies: null
    }
  }

  async _fetchExpressionDataForHeatmap({resource, host}) {
    this.setState({
      isLoading: true
    })

    const url = URI(resource, host).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      const { species } = this.props

      let results = Object.values(await response.json())[0][species]

      const experimentAccessions = Object.keys(results)
      const markerGenes = Object.keys(results[experimentAccessions[0]].markerGeneExpression).sort()
      const technologyType = Object.values(results).map(experimentAccession => experimentAccession.technologyType)

      this.setState({
        expressionByCellTypeData: response,
        responseBySpecies: results,
        axisData: this._heatmapAxis(markerGenes, results, technologyType),
        heatmapData: this._factorHeatmapData(markerGenes, results),
        filteredData: this._factorHeatmapData(markerGenes, results),
        isLoading: false,
        selectedSpecies: null
      })

    } catch(e) {
      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  _factorHeatmapData(markerGenesBySpecies, responseBySpecies) {

    const experimentAccessions = Object.keys(responseBySpecies)

    // const cellTypes = experimentAccessions.map(accession =>
    //   Object.keys(responseBySpecies[accession][Object.keys(responseBySpecies[accession])[0]])
    // )

    const cellTypes = Object.keys(Object.values((Object.values(responseBySpecies)[0]).markerGeneExpression)[0])

    const heatmapData = markerGenesBySpecies.map(markerGene =>
      _.flatMap(
        experimentAccessions.map((experimentAccession) =>
          cellTypes.map(cellType => {
            return (responseBySpecies[experimentAccession]).markerGeneExpression[markerGene][cellType]
          })
        )
      )
    )
    return heatmapData
  }

  _heatmapAxis(markerGenesBySpecies, responseBySpecies, technologyType) {
    const experimentAccessions = Object.keys(responseBySpecies)
    const cellType = Object.keys(Object.values((Object.values(responseBySpecies)[0]).markerGeneExpression)[0])[0]

    const yAxisCategory = _.flatMap(
      experimentAccessions.map((experimentAccession, idx) =>
        ({
          cellType: cellType.substring(1, cellType.length - 1),
          experimentAccession: experimentAccession,
          technologyType: technologyType[idx][0]
        })
      )
    )
    return ({ y: markerGenesBySpecies, x: yAxisCategory })
  }

  componentDidMount() {
    this._fetchExpressionDataForHeatmap(this.props)
  }

  componentDidUpdate(previousProps) {
    if (previousProps.resource !== this.props.resource) {
      this._fetchExpressionDataForHeatmap(this.props)
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

  render() {
    const { axisData, hasError, isLoading, filteredData } = this.state
    const { wrapperClassName, plotWrapperClassName, species } = this.props
    const { hasDynamicHeight, defaultHeatmapHeight, heatmapRowHeight } = this.props

    return (
      hasError ?
        <CalloutAlert error={hasError}/> :

        <div className={wrapperClassName}>
          <div className={plotWrapperClassName} style={{position: `relative`}}>
            <CellTypeHighchartsHeatmap
              species={{species}}
              axisData={axisData}
              heatmapData={filteredData}
              chartHeight={defaultHeatmapHeight}
              hasDynamicHeight={axisData.y && axisData.y.length > 5 ? hasDynamicHeight : false} // don't want dynamic height if there is little or no data
              heatmapRowHeight={heatmapRowHeight}
            />
            <LoadingOverlay
              show={isLoading}
            />
          </div>
          <DataSeriesHeatmapLegent />
        </div>
    )
  }
}

CellTypeView.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,
  plotWrapperClassName: PropTypes.string,
  defaultHeatmapHeight: PropTypes.number,
  hasDynamicHeight: PropTypes.bool,
  heatmapRowHeight: PropTypes.number,
  species: PropTypes.string.isRequired
}

CellTypeView.defaultProps = {
  wrapperClassName: `small-8 columns row`,
  plotWrapperClassName: `small-12 columns`,
  defaultHeatmapHeight: 300,
  hasDynamicHeight: true,
  heatmapRowHeight: 20
}

export default CellTypeView
