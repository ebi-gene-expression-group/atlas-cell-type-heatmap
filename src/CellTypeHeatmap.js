import React from "react"
import PropTypes from 'prop-types'

import URI from 'urijs'
import _ from "lodash"

import HighchartsHeatmap from "./HighchartsHeatmap"
import CalloutAlert from './CalloutAlert'
import LoadingOverlay from './LoadingOverlay'
import PlotSettingsDropdown from './PlotSettingsDropdown'


class CellTypeHeatmap extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      expressionByCellTypeData: [],
      solrResponse: [],
      experimentAccessions: [],
      responseByAccession: [],
      axisData: {},
      selectedExperiment: null
    }
  }

  async _fetchExpressionDataForHeatmap({resource, host}) {
    this.setState({
      isLoading: true
    })

    //  `http://localhost:8080/gxa/sc/json/experiments/celltype/name/organism_part/value/pancreas`
    // "https://gist.githubusercontent.com/lingyun1010/095f414db81d34cef2a3bc8eaf56544f/raw/e898dd2ce08fb427e929df6c8dac262920f244a7/cellTypeAPIJsonResponse.json"

    const url = URI(resource, host).toString()

    try {
      const response = await fetch(url)

      if (!response.ok) {
        throw new Error(`${url} => ${response.status}`)
      }

      const results = Object.values(await response.json())[0]
      const experimentAccessions = Object.keys(results)
      const markerGenes = Object.keys(results[experimentAccessions[0]]).sort()

      this.setState({
        markerGenes: markerGenes,
        experimentAccessions: Object.keys(results),
        expressionByCellTypeData: response,
        responseByAccession: results,
        axisData: this._heatmapAxis(markerGenes, results),
        heatmapData: this._factorHeatmapData(markerGenes, results),
        filteredData: this._factorHeatmapData(markerGenes, results),
        isLoading: false,
        selectedExperiment: null
      })

    } catch(e) {
      this.setState({
        data: null,
        isLoading: false,
        hasError: {
          description: `There was a problem communicating with the server. Please try again later.`,
          name: e.name,
          message: e.message
        }
      })
    }
  }

  _factorHeatmapData(markerGenes, results, selectedExperiment) {
    const experimentAccessions = selectedExperiment ? [selectedExperiment] : Object.keys(results)
    const cellTypes = experimentAccessions.map(accession =>
      Object.keys(results[accession][Object.keys(results[accession])[0]])
    )
    const heatmapData = markerGenes.map(markerGene =>
      _.flatMap(
        experimentAccessions.map((experimentAccession, idx) =>
          cellTypes[idx].map(cellType => {
            return results[experimentAccession][markerGene][cellType]
          })
        )
      )
    )
    return heatmapData
  }

  _heatmapAxis(markerGenes, results, selectedExperiment) {
    const experimentAccessions = selectedExperiment ? [selectedExperiment] : Object.keys(results)
    const cellTypes = experimentAccessions.map(accession =>
      Object.keys(results[accession][Object.keys(results[accession])[0]])
    )

    const yAxisCategory = _.flatMap(
      experimentAccessions.map((experimentAccession, idx) =>
        cellTypes[idx].map(
          cellType => `${cellType} <br> ${experimentAccession}`
        )
      )
    )
    return ({ y: markerGenes, x: yAxisCategory })
  }

  componentDidMount() {
    this._fetchExpressionDataForHeatmap(this.props)
  }

  componentDidUpdate(previousProps) {
    if (previousProps.resource !== this.props.resource) {
      this._fetchExpressionDataForHeatmap(this.props)
    }
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: {
        description: `There was a problem rendering this component.`,
        name: error.name,
        message: `${error.message} â€“ ${info}`
      }
    })
  }

  render() {
    const { heatmapData, axisData, hasError, isLoading } = this.state
    const { wrapperClassName, plotWrapperClassName } = this.props
    const { hasDynamicHeight, defaultHeatmapHeight, heatmapRowHeight } = this.props
    const { selectedExperiment, experimentAccessions, markerGenes, responseByAccession, filteredData } = this.state

    const experimentOptions = experimentAccessions
      .sort((a, b) => a-b)
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    if (experimentOptions.length > 1) {
      experimentOptions.unshift({
        value: `all`,
        label: `All`,
        isDisabled: false
      })
    }

    return (
      hasError ?
        <CalloutAlert error={hasError}/> :
        <div>
          <div className={wrapperClassName}>
            <div className={`small-12 medium-6 columns`}>
              <PlotSettingsDropdown
                labelText={`Show Gene Expressions for Experiment:`}
                options={experimentOptions}
                onSelect={(selectedOption) =>{
                  this.setState({
                    filteredData: selectedOption.value === `all` ?
                      _.cloneDeep(heatmapData) :
                      this._factorHeatmapData(markerGenes, responseByAccession, selectedOption.value),
                    axisData: selectedOption.value === `all` ?
                      this._heatmapAxis(markerGenes, responseByAccession) :
                      this._heatmapAxis(markerGenes, responseByAccession, selectedOption.value),
                    selectedExperiment: selectedOption
                  })
                }}
                value={selectedExperiment || experimentOptions[0]}
              />
            </div>
          </div>
          <div className={wrapperClassName}>
            <div className={plotWrapperClassName} style={{position: `relative`}}>
              <HighchartsHeatmap
                axisData={axisData}
                heatmapData={filteredData}
                chartHeight={defaultHeatmapHeight}
                hasDynamicHeight={axisData.y && axisData.y.length > 5 ? hasDynamicHeight : false} // don't want dynamic height if there is little or no data
                heatmapRowHeight={heatmapRowHeight}
              />
              <LoadingOverlay
                show={isLoading}
              />
            </div>
          </div>
        </div>
    )
  }
}

CellTypeHeatmap.propTypes = {
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,
  plotWrapperClassName: PropTypes.string,
  defaultHeatmapHeight: PropTypes.number,
  hasDynamicHeight: PropTypes.bool,
  heatmapRowHeight: PropTypes.number
}

CellTypeHeatmap.defaultProps = {
  wrapperClassName: `row`,
  plotWrapperClassName: `small-12 columns`,
  defaultHeatmapHeight: 300,
  hasDynamicHeight: true,
  heatmapRowHeight: 20
}
export default CellTypeHeatmap
