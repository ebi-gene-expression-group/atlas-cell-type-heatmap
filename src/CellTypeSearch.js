import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'
import CellTypeView from './CellTypeView'
import LoadingOverlay from "./LoadingOverlay"
import PlotSettingsDropdown from "./PlotSettingsDropdown"
import _ from "lodash"

class CellTypeSearch extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      selectedSpecies: ``,
      selectedCellType: ``,
      cellType: ``
    }
  }

  render() {
    const requestParams = URI(location.search).query(true)
    const {speciesList, cellTypePayload, host, resource} = this.props
    const {selectedSpecies, selectedCellType, isLoading} = this.state

    const cellTypeList = speciesList.map(species => _.uniq(_.flatMap(Object.values(cellTypePayload[species]))))

    const cellTypesOptions = selectedSpecies ?
      cellTypeList[speciesList.indexOf(selectedSpecies)]
        .map((v) => ({
          value: v.toString().substring(1, v.length - 1),
          label: v.toString().substring(1, v.length - 1),
          isDisabled: false
        }))
        .filter(v => v.value !== `not available` && v.value !== `not applicable` && v.value !== ``) :
      []

    const speciesOptions = speciesList
      .map((v) => ({
        value: v.toString(),
        label: v,
        isDisabled: false
      }))

    return (
      <div>
        <div className={`row expanded`}>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Species:`}
              options={speciesOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedSpecies: selectedOption.value
                })
              }}
              value={{value: selectedSpecies || requestParams.species, label: selectedSpecies || requestParams.species}}
            />
          </div>
          <div className={`small-12 medium-5 columns`}>
            <PlotSettingsDropdown
              labelText={`Inferred cell types:`}
              options={cellTypesOptions}
              onSelect={(selectedOption) => {
                this.setState({
                  selectedCellType: selectedOption.value
                })
              }}
              value={{value: selectedCellType || requestParams.cellType, label: selectedCellType || requestParams.cellType}}
            />
          </div>
          <div className={`small-12 medium-2 columns`} style={{paddingTop: `25px`}}>
            <button className={`button`} onClick={(event) => {
              event.preventDefault()
              this.props.history.push(`/search?species=` + this.state.selectedSpecies + `&cellType=` + this.state.selectedCellType)
              this.setState({
                cellType: this.state.selectedCellType
              })}}>
              Search
            </button>
          </div>

        </div>
        {
          requestParams.cellType && <div style={{paddingBottom: `25px`}}>
            <CellTypeView
              wrapperClassName={`row expanded`}
              resource={`${resource}/${requestParams.cellType}`}
              host={host}
              hasDynamicHeight={true}
              species={requestParams.species}
              heatmapRowHeight={30}
            />
          </div>
        }
        <LoadingOverlay
          show={isLoading}
        />
      </div>
    )
  }
}

CellTypeSearch.propTypes = {
  history: PropTypes.object.isRequired,
  host: PropTypes.string.isRequired,
  resource: PropTypes.string.isRequired,
  speciesList: PropTypes.arrayOf(PropTypes.string).isRequired,
  cellTypeList: PropTypes.arrayOf(PropTypes.string).isRequired
}

export default CellTypeSearch