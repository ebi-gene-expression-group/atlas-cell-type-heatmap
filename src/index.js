import React from 'react'
import { BrowserRouter, Route } from 'react-router-dom'

import CellTypeSearch from './CellTypeSearch'

const CellTypeHeatmapRouter = (options) =>
  <BrowserRouter basename={``}>
    <Route
      path={`/`}
      render={
        ({history}) =>
          <CellTypeSearch
            {...options}
            history={history}/>
      } />
  </BrowserRouter>


export default CellTypeHeatmapRouter