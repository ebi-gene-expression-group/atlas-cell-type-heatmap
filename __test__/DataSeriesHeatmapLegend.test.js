import React from 'react'
import renderer from 'react-test-renderer'

import Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import DataSeriesHeatmapLegend from '../src/DataSeriesHeatmapLegend'
import { DataSeriesHeatmapLegendBox } from '../src/DataSeriesHeatmapLegend'

Enzyme.configure({ adapter: new Adapter() })

describe(`PlotSettingsDropdown`, () => {
  test(`contains four coloured legend boxes with default props`, () => {
    const wrapper = shallow(<DataSeriesHeatmapLegend />)
    expect(wrapper.find(DataSeriesHeatmapLegendBox).length).toBe(4)
  })

  test(`with no explicit props matches snapshot`, () => {
    const tree = renderer
      .create(<DataSeriesHeatmapLegend />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
