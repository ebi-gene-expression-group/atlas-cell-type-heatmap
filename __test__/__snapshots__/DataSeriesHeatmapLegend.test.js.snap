// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlotSettingsDropdown with no explicit props matches snapshot 1`] = `
<div
  className="sc-bxivhb bDTNtS"
>
  <div
    className="sc-EHOje cizzKT"
    color="white"
  >
    C
  </div>
  CPM
  <div
    className="sc-EHOje ccSpQn"
    color="white"
  >
    T
  </div>
  TPM
  <br />
  <div
    className="sc-bdVaJa XCEfz"
  >
    <span
      className="sc-ifAKCX iySXmL"
      data-icon="i"
      data-placement="bottom"
      data-toggle="tooltip"
      title="Single cell expression CPM. Low: <10, Medium: 10-1,000, High: >1,000. "
    />
  </div>
  <div
    className="sc-bdVaJa XCEfz"
  >
    <div
      className="sc-bwzfXH buNHcc"
    />
    <span
      className="sc-htpNat jjecFx"
    >
      High
    </span>
  </div>
  <div
    className="sc-bdVaJa XCEfz"
  >
    <div
      className="sc-bwzfXH efQDwT"
    />
    <span
      className="sc-htpNat jjecFx"
    >
      Medium
    </span>
  </div>
  <div
    className="sc-bdVaJa XCEfz"
  >
    <div
      className="sc-bwzfXH beksOx"
    />
    <span
      className="sc-htpNat jjecFx"
    >
      Low
    </span>
  </div>
  <div
    className="sc-bdVaJa XCEfz"
  >
    <div
      className="sc-bwzfXH kpsXFF"
    />
    <span
      className="sc-htpNat jjecFx"
    >
      No data available
    </span>
  </div>
</div>
`;
